From 1ec9148708a4b1cd32d630ad90d2dcd0949873cb Mon Sep 17 00:00:00 2001
From: bwetherfield <bwetherfield@gmail.com>
Date: Sun, 4 Nov 2018 01:37:53 -0400
Subject: [PATCH 35/66] Graph schemes (#156)

* Add weightScheme and adjacencyScheme constructors

* Move weightScheme pullback

* Move adjacencyScheme pullback

* Make scheme protocols public

* Add weightScheme initializer to WeightedGraphProtocol

* Add weightedScheme constructor

* Add public permission to Carrying structs
---
 .../Wetherfield/AdjacencyCarrying.swift       |  2 +-
 .../DirectedGraphSchemeProtocol.swift         |  2 +-
 .../Wetherfield/GraphSchemeProtocol.swift     |  2 +-
 .../UndirectedGraphSchemeProtocol.swift       |  2 +-
 .../UnweightedGraphSchemeProtocol.swift       |  2 +-
 .../Wetherfield/WeightCarrying.swift          | 13 +++++++++++-
 .../WeightedGraphProtocolExtension.swift      | 20 +++++++++++++++++++
 .../WeightedGraphSchemeProtocol.swift         |  2 +-
 8 files changed, 38 insertions(+), 7 deletions(-)
 create mode 100644 Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphProtocolExtension.swift

diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/AdjacencyCarrying.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/AdjacencyCarrying.swift
index c3c32ca..d0a56c8 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/AdjacencyCarrying.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/AdjacencyCarrying.swift
@@ -7,7 +7,7 @@
 
 import DataStructures
 
-struct AdjacencyCarrying <G: GraphProtocol> {
+public struct AdjacencyCarrying <G: GraphProtocol> {
     let contains: (G.Edge) -> Bool
     
     private init (contains: @escaping (G.Edge) -> Bool) {
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/DirectedGraphSchemeProtocol.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/DirectedGraphSchemeProtocol.swift
index a0814cf..955f7c2 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/DirectedGraphSchemeProtocol.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/DirectedGraphSchemeProtocol.swift
@@ -7,4 +7,4 @@
 
 import DataStructures
 
-protocol DirectedGraphSchemeProtocol: GraphSchemeProtocol where Edge == OrderedPair<Node> { }
+public protocol DirectedGraphSchemeProtocol: GraphSchemeProtocol where Edge == OrderedPair<Node> { }
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/GraphSchemeProtocol.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/GraphSchemeProtocol.swift
index ad11593..2b3a6c1 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/GraphSchemeProtocol.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/GraphSchemeProtocol.swift
@@ -7,7 +7,7 @@
 
 import DataStructures
 
-protocol GraphSchemeProtocol {
+public protocol GraphSchemeProtocol {
     associatedtype Node
     associatedtype Edge: SymmetricPair where Edge.A == Node
 }
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/UndirectedGraphSchemeProtocol.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/UndirectedGraphSchemeProtocol.swift
index 1184b5f..e6d1520 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/UndirectedGraphSchemeProtocol.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/UndirectedGraphSchemeProtocol.swift
@@ -7,4 +7,4 @@
 
 import DataStructures
 
-protocol UndirectedGraphSchemeProtocol: GraphSchemeProtocol where Edge == UnorderedPair<Node> { }
+public protocol UndirectedGraphSchemeProtocol: GraphSchemeProtocol where Edge == UnorderedPair<Node> { }
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/UnweightedGraphSchemeProtocol.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/UnweightedGraphSchemeProtocol.swift
index 4cea885..83bf759 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/UnweightedGraphSchemeProtocol.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/UnweightedGraphSchemeProtocol.swift
@@ -7,7 +7,7 @@
 
 import DataStructures
 
-protocol UnweightedGraphSchemeProtocol: GraphSchemeProtocol {
+public protocol UnweightedGraphSchemeProtocol: GraphSchemeProtocol {
     var contains: (Edge) -> Bool { get }
     
     init (_ contains: @escaping (Edge) -> Bool)
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightCarrying.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightCarrying.swift
index ecbf260..8a11396 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightCarrying.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightCarrying.swift
@@ -7,7 +7,7 @@
 
 import DataStructures
 
-struct WeightCarrying<G: WeightedGraphProtocol> {
+public struct WeightCarrying<G: WeightedGraphProtocol> {
     let weight: (G.Edge) -> G.Weight?
     
     private init (weight: @escaping (G.Edge) -> G.Weight?) {
@@ -73,6 +73,17 @@ extension WeightCarrying {
     }
 }
 
+extension WeightCarrying {
+    
+    func weightedScheme <Scheme: WeightedGraphSchemeProtocol> () -> Scheme where
+        Scheme.Edge == G.Edge,
+        Scheme.Node == G.Node,
+        Scheme.Weight == G.Weight
+    {
+        return Scheme.init(weight)
+    }
+}
+
 extension WeightCarrying where G: DirectedGraphProtocol {
     
     static func * <H> (lhs: WeightCarrying, rhs: WeightCarrying<H>) -> WeightCarrying
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphProtocolExtension.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphProtocolExtension.swift
new file mode 100644
index 0000000..4ea02d7
--- /dev/null
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphProtocolExtension.swift
@@ -0,0 +1,20 @@
+//
+//  WeightedGraphProtocolExtension.swift
+//  SpelledPitch
+//
+//  Created by Benjamin Wetherfield on 03/11/2018.
+//
+
+import DataStructures
+
+extension WeightedGraphProtocol {
+    @inlinable
+    func weightScheme <G> () -> G where
+        G: WeightedGraphSchemeProtocol,
+        G.Edge == Edge,
+        G.Weight == Weight,
+        G.Node == Node
+    {
+        return G.init(self.weight)
+    }
+}
diff --git a/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphSchemeProtocol.swift b/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphSchemeProtocol.swift
index 13a6a5e..dce5b8b 100644
--- a/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphSchemeProtocol.swift
+++ b/Sources/SpelledPitch/PitchSpeller/Wetherfield/WeightedGraphSchemeProtocol.swift
@@ -7,7 +7,7 @@
 
 import DataStructures
 
-protocol WeightedGraphSchemeProtocol: GraphSchemeProtocol {
+public protocol WeightedGraphSchemeProtocol: GraphSchemeProtocol {
     associatedtype Weight
     
     var weight: (Edge) -> Weight? { get }
-- 
2.17.2 (Apple Git-113)

